grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy

```diff
@@ -338,77 +336,30 @@ class StockMovementApiController {
-                    throw new IllegalArgumentException("Requisition item id: ${requisitionItemId} not found")
-                }
-
-                // FIXME Should find bin location by name and parent and inventory item by lot number and expiration date
-                // and compare object equality (or at least PK equality) rather than comparing various components
-                AvailableItem availableItem = pickPageItem.availableItems?.find {
-                    (binLocation ? it.binLocation?.name == binLocation : !it.binLocation) && lotNumber == (it.inventoryItem?.lotNumber ?: null) &&
-                            expirationDate == (it?.inventoryItem?.expirationDate ? it.inventoryItem.expirationDate.format(Constants.EXPIRATION_DATE_FORMAT) : null)
-                }
+        MultipartFile importFile = command.importFile
+        if (importFile.isEmpty()) {
+            throw new IllegalArgumentException("File cannot be empty")
+        }
 
-                if (!availableItem) {
-                    throw new IllegalArgumentException("There is no item available with lot: ${lotNumber ?: ""}, expiration date: ${tokens[2] ?: ""} and bin: ${binLocation ?: ""}")
-                }
+        String csv = new String(importFile.bytes)
+        List<Map> importedLines = stockMovementService.parsePickCsvTemplateImport(csv)
 
-                RequisitionItem requisitionItem = pickPageItem.requisitionItem?.modificationItem ?: pickPageItem.requisitionItem
+        stockMovementService.validatePicklistListImport(importedLines, pickPageItems)
 
-                pickPageItem.picklistItems.each {
-                    if (it.id) {
-                        it.quantity = 0
-                    }
-                }
-                pickPageItem.picklistItems.add(new PicklistItem(
-                        requisitionItem: requisitionItem,
-                        inventoryItem: availableItem.inventoryItem,
-                        binLocation: availableItem.binLocation,
-                        quantity: quantityPicked,
-                        sortOrder: pickPageItem.sortOrder
-                ))
-            }
+        List errors = importedLines*.errors.withIndex().collect { errors, index ->
+            errors.collect { "Row ${1}: ${it}" }
+        }.flatten()
 
-            stockMovementService.createOrUpdatePicklistItem(stockMovement, pickPageItems)
+        stockMovementService.importPicklistTemplate(importedLines, stockMovement, pickPageItems)
 
-        } catch (Exception e) {
-            // FIXME The global error handler does not return JSON for multipart uploads
-            log.warn("Error occurred while importing CSV: " + e.message, e)
-            response.status = 500
-            render([errorCode: 500, errorMessage: e?.message ?: "An unknown error occurred during import"] as JSON)
-            return
+        if (!errors.isEmpty()) {
+            render([message: "Data imported with errors", errors: errors] as JSON)
```

Comment:
:astonished: I guess this answers my first comment, but I'm in shock if we really want to go this path. 
This smells like an anti-pattern to me, that even though there are errors, we would get 200 status and act like everything went well.
P.S. I know this was not your choice to do it this way, but wanted to raise my astonishment.

Reply:
> I guess this answers my first comment, but I'm in shock if we really want to go this path.
> This smells like an anti-pattern to me, that even though there are errors, we would get 200 status and act like everything went well.
> P.S. I know this was not your choice to do it this way, but wanted to raise my astonishment.

@kchelstowski Can you elaborate on this? I feel like I'm missing some context here.

Reply:
@jmiranda this comment was about importing the template with e.g. 50 valid lines and 50 invalid - I was just surprised that we would proceed even if we were about to save 50 items out of 100. 
This is why I was surprised that we first try to import the items and then eventually check if the errors are not empty - in typical scenario we do the opposite.